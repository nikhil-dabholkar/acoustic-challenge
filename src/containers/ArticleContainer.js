import React from 'react';
import { connect } from 'react-redux';
import { fetchSelectedArticleData, fetchAllArticlesData } from '../actionCreators/ArticleActions';
import { switchLanguagePreference } from '../actionCreators/LanguageActions';
import { BrowserRouter as Router, Route, Switch, Redirect, Link } from "react-router-dom";
import { lazy, Suspense } from 'react';
import { LoadingComponent, LanguageSwitcherComponent } from '../components/index';
import { ErrorBoundaryWrapper } from '../wrappers/index';
import { Breadcrumb, BreadcrumbItem } from 'carbon-components-react';
import styles from './styles/ArticleContainer.module.scss';
import HeaderComponent from '../components/shared/HeaderComponent';
const ListComponent = lazy(() => import("../components/article/ListComponent"));
const DetailsComponent = lazy(() => import("../components/article/DetailsComponent"));
const ErrorComponent = lazy(() => import("../components/shared/ErrorComponent"));

function ArticleContainer({ language, switchLanguage, fetchArticle, fetchAllArticles, article }) {
    return (<React.Fragment>
        <Router>
            <HeaderComponent language={language} switchLanguage={switchLanguage}></HeaderComponent>
            <ErrorBoundaryWrapper>
                <Switch>
                    <Route path="/articles" render={(props) => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <ListComponent loading={article.loading} error={article.error} allArticles={article.allArticles} fetchAllArticles={fetchAllArticles} {...props} />
                        </Suspense>);
                    }} exact />
                    <Route path="/articles/:id" render={(props) => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <DetailsComponent loading={article.loading} error={article.error} selectedArticle={article.selectedArticle} fetchArticle={fetchArticle} {...props} />
                        </Suspense>)
                    }} exact />
                    <Route path="/" render={() => <Redirect to="/articles" />} exact></Route>
                    <Route path="*" render={() => {
                        return (<Suspense fallback={<LoadingComponent />}>
                            <ErrorComponent />
                        </Suspense>)
                    }} exact />
                </Switch>
            </ErrorBoundaryWrapper>
        </Router>

    </React.Fragment>
    );
}

const mapStateToProps = state => {
    return ({
        article: state.articleReducer,
        language: state.languageReducer
    })
};

const mapDispatchToProps = {
    fetchArticle: fetchSelectedArticleData,
    fetchAllArticles: fetchAllArticlesData,
    switchLanguage: switchLanguagePreference
}

export default connect(mapStateToProps, mapDispatchToProps)(ArticleContainer);
