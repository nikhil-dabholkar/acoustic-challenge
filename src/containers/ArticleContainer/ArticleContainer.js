import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchSelectedArticleData, fetchAllArticlesData, resetArticlesError } from '../../actionCreators/ArticleActions';
import { Route, useLocation } from "react-router-dom";
import { useHistory } from 'react-router-dom'
import { lazy, Suspense } from 'react';
import { LoadingComponent } from '../../components/index';
import { ErrorBoundaryWrapper } from '../../wrappers/index';

const ListComponent = lazy(() => import("../../components/article/ListComponent/ListComponent"));
const DetailsComponent = lazy(() => import("../../components/article/DetailsComponent/DetailsComponent"));

function ArticleContainer({ resetError, fetchArticle, fetchAllArticles, article }) {
    const history = useHistory();
    const location = useLocation();
    const errorBoundaryRef = useRef(null);

    useEffect(() => {
        /**
         * When ever the url changes, clear the error 
         */
        return history.listen((location) => {
            resetError();
        })
    }, [history])

    return (<React.Fragment>
        <ErrorBoundaryWrapper key={location.pathname} ref={errorBoundaryRef}>
            <div data-testid="articleContainer">
                <Route path="/articles" render={(props) => {
                    return (<Suspense fallback={<LoadingComponent />}>
                        <ListComponent loading={article.loading} error={article.error} allArticles={article.allArticles} fetchAllArticles={fetchAllArticles} {...props} />
                    </Suspense>);
                }} exact />
                <Route path="/articles/:id" render={(props) => {
                    return (<Suspense fallback={<LoadingComponent />}>
                        <DetailsComponent loading={article.loading} error={article.error} selectedArticle={article.selectedArticle} fetchArticle={fetchArticle} {...props} />
                    </Suspense>)
                }} exact />
            </div>
        </ErrorBoundaryWrapper>
    </React.Fragment>
    );
}

const mapStateToProps = state => {
    return ({
        article: state.articleReducer
    })
};

const mapDispatchToProps = {
    fetchArticle: fetchSelectedArticleData,
    fetchAllArticles: fetchAllArticlesData,
    resetError: resetArticlesError
}

ArticleContainer.propTypes = {
    resetError: PropTypes.func.isRequired,
    fetchArticle: PropTypes.func.isRequired,
    fetchAllArticles: PropTypes.func.isRequired,
    article: PropTypes.object.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(React.memo(ArticleContainer));
