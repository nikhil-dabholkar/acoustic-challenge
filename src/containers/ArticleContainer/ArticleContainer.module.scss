.header {
    padding: 15px;
    overflow: auto;
    font-family: "IBM Plex Sans VF", "Helvetica Neue", Arial, sans-serif;
    .logoContainer {
        font-size: 30px;
        font-weight: 700;
        padding: 0 20px;
        float: left;
    }
    .languageSwitcherContainer {
        float: right;
        padding: 0 20px;
    }
}
