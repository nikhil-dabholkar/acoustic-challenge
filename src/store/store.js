import rootReducer from '../reducer/index';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';

export const store = createStore(rootReducer, applyMiddleware(thunk))