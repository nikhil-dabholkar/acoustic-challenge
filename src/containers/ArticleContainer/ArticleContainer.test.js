import React from 'react'
import ArticleContainer from './ArticleContainer'
import { IntlProvider } from 'react-intl'
import { store } from '../../store/store'
import { Provider } from 'react-redux'
import { render, cleanup, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import enUS from '../../i18n/en_US.json'

afterEach(cleanup)

function renderWithRouter (
  ui,
  { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  return {
    ...render(<Router history={history}>{ui}</Router>),
    history
  }
}

function rendererWithRouter (
  ui,
  { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  return {
    ...renderer.create(<Router history={history}>{ui}</Router>),
    history
  }
}

describe('ArticleContainer', () => {
  test('Check if ArticleContainer Renders on /articles/fa9519d5-0363-4b8d-8e1f-627d802c08a8', async () => {
    const route = '/articles/fa9519d5-0363-4b8d-8e1f-627d802c08a8'
    renderWithRouter(<Provider store={store}><IntlProvider locale="en" messages={enUS}><ArticleContainer></ArticleContainer></IntlProvider></Provider>, { route })
    const lazyElement = await screen.findByTestId('articleContainer')
    expect(lazyElement).toBeInTheDocument()
  })

  test('Check if ArticleContainer Renders on /articles', async () => {
    const route = '/articles'
    renderWithRouter(<Provider store={store}><IntlProvider locale="en" messages={enUS}><ArticleContainer></ArticleContainer></IntlProvider></Provider>, { route })
    const lazyElement = await screen.findByTestId('listComponent')
    expect(lazyElement).toBeInTheDocument()
  })

  test('Check if matches ArticleContainer Snapshot', async () => {
    const route = '/articles'
    const tree = rendererWithRouter(<Provider store={store}><IntlProvider locale="en" messages={enUS}><ArticleContainer></ArticleContainer></IntlProvider></Provider>, { route }).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
