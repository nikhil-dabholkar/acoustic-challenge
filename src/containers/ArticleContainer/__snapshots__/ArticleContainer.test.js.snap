// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleContainer Check if matches ArticleContainer Snapshot 1`] = `
<div
  data-testid="articleContainer"
>
  <div
    className="container"
    data-testid="listComponent"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="ag-theme-alpine listComponentContainer textAlignLeft"
        >
          <div
            style={
              Object {
                "height": "100%",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
